lib/src/sources/dmi_source.cc

```diff
@@ -0,0 +1,207 @@
+#include <leatherman/file_util/file.hpp>
+#include <leatherman/execution/execution.hpp>
+#include <boost/filesystem.hpp>
+#include <boost/algorithm/string.hpp>
+
+using namespace std;
+using namespace boost::filesystem;
+namespace bs = boost::system;
+namespace lth_file = leatherman::file_util;
+using namespace leatherman::util;
+
+namespace whereami { namespace sources {
+
+    std::string dmi_base::bios_vendor() const
+    {
+        return data_ ? data_->bios_vendor : "";
+    }
+
+    std::string dmi_base::board_manufacturer() const
+    {
+        return data_ ? data_->board_manufacturer : "";
+    }
+
+    std::string dmi_base::board_product_name() const
+    {
+        return data_ ? data_->board_product_name : "";
+    }
+
+    std::string dmi_base::product_name() const
+    {
+        return data_ ? data_->product_name : "";
+    }
+
+    std::string dmi_base::manufacturer() const
+    {
+        return data_ ? data_->manufacturer : "";
+    }
+
+    dmi::dmi()
+    {
+        collect_data();
+    }
+
+    void dmi::collect_data()
+    {
+        if (!collect_sys_data()) {
+            collect_dmidecode_data();
+        }
+    }
+
+    string dmi::sys_path(string const& filename) const
+    {
+        return "/sys/class/dmi/id/" + filename;
+    }
+
+    bool dmi::collect_sys_data()
+    {
+        // Check that /sys/class/dmi exists
+        bs::error_code ec;
+        if (is_directory(sys_path(), ec)) {
+            data_.reset(new dmi_data);
+            data_->bios_vendor = read_file(sys_path("bios_vendor"));
+            data_->board_manufacturer = read_file(sys_path("board_vendor"));
+            data_->board_product_name = read_file(sys_path("board_name"));
+            data_->manufacturer = read_file(sys_path("sys_vendor"));
+            data_->product_name = read_file(sys_path("product_name"));
+        } else {
+            LOG_DEBUG(sys_path() + " cannot be accessed.");
+        }
+        return data_.get() != nullptr;
+    }
+
+    bool dmi::collect_dmidecode_data()
+    {
+        LOG_DEBUG("Using dmidecode to query DMI information.");
+
+        string dmidecode = leatherman::execution::which("dmidecode");
+
+        if (dmidecode.empty()) {
+            LOG_DEBUG("dmidecode executable not found");
+        } else {
+            int dmi_type {-1};
+            leatherman::execution::each_line(dmidecode, [&](string& line) {
+                parse_dmidecode_line(line, dmi_type);
+                return true;
+            });
+        }
+        return data_.get() != nullptr;
+    }
+
+    // Read a single file in the DMI directory
+    string dmi::read_file(string const& path)
+    {
+        bs::error_code ec;
+        if (!is_regular_file(path, ec)) {
+            LOG_DEBUG("{1}: {2}.", path, ec.message());
+            return {};
+        }
+
+        string value;
+        if (!lth_file::read(path, value)) {
+            LOG_DEBUG("{1}: file could not be read.", path);
+            return {};
```

Comment:
curly braces mean 'create a default value of the type', which for a string is an empty one